### PR TITLE
GH-2392 elasticsearch upgrade

### PR DESCRIPTION
**Pull request by barthanssens**
GitHub issue resolved: #2392 

Briefly describe the changes proposed in this PR:

- upgrade Elasticsearch / Lucene libraries
- upgrade to Solr 8.4 (instead of 8.5, since 8.5 version seems to break some tests,  which might be related to highlighting)
- use ElasticsearchClusterRunner instead of EmbeddedElasticsearch (which doesn't support ES7) and updated tests accordingly
- raise the Xmx to 2 GB (needed for ElasticsearchClusterrunner)
- remove explicit snakeyaml dependency from pom.xml (introduced when moving to ES6.8)
- excluded jetty-alpn HTTP/2, since it caused problems (see also https://github.com/jetty-project/jetty-alpn/issues/15)
- use ES seqNo en primaryTerm instead of (deprecated) document Version to detect concurrent update issues (see also https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking-changes-7.0.html#_internal_versioning_is_no_longer_supported_for_optimistic_concurrency_control)
- reindex lucene index when the index was written by an older version 

---- 
PR Author Checklist (see the [contributor guidelines](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md) for more details):

 - [x] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [x] I've added tests for the changes I made
 - [x] I've applied [code formatting](https://github.com/eclipse/rdf4j/blob/master/.github/CONTRIBUTING.md#code-formatting) (you can use `mvn process-resources` to format from the command line)
 - [x] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [x] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by barthanssens** [comment]
Still have to check this change (ES 7 allows to either get the exact number of results, or just a fast "at least 10.000")  

https://www.elastic.co/guide/en/elasticsearch/reference/7.8/search-request-body.html#request-body-search-track-total-hits

**Comment by barthanssens** [comment]
Another useful reference / command line tool for upgrading Lucene indexes: https://lucene.apache.org/core/8_5_0/core/org/apache/lucene/index/IndexUpgrader.html

**Comment by barthanssens** [comment]
Some quick feedback on this draft PR would be appreciated :-) 

As mentioned on gitter: setting the JVM in the benchmark is currently ignored since it's not supported by ElasticsearchClusterRunner, and it doesn't seem portable ?

And/or ideas for upgrade path: 
- Lucene has a command line tool for reindexing, so I guess that's covered
- Solr search: assuming a v8-client can contact older servers, this should be ok (to be tested)  ?
- ES search: same as Solr ?
- ES store: that's going to be the hardest part (the ES docs do mention a manual upgrade procedure, but that seems a bit cumbersome...)

**Comment by barthanssens** [comment]
Well, I haven't filed CQs yet - I'd be happy to do so - because I wanted to be a bit more certain about the versions first :-)

Jetty is already used by the workbench/server and an Eclipse project, so in terms of CQs, that's already covered. I did have to exclude the ALPN (used for TLS over HTTP/2 connections) because it tripped some enforcer rules, though perhaps I should look into it... 

Not sure if excluding Netty would work... At first glance there are quite a few CQs for other projects listed in IPZilla, and under the new Eclipse rules there is probably no need for filing piggyback-CQs, so that would reduce the administrative load a little bit.

**Comment by hmottestad** [comment]
Just to comment on setting the jvm in the benchmarks. I found we needed to do this to be able to run the benchmarks. It wasn't very portable no, but it was the only way I could find to make it work.

**Comment by barthanssens** [comment]
@hmottestad  perhaps these settings could be moved to a new profile in a pom.xml ? 
Would make it a bit more portable IMHO, but I'm not familiar with the benchmarking, so I don't know if that'll work for you...

**Comment by hmottestad** [comment]
I run the benchmarks directly from IntelliJ with a JMH plugin someone made.

**Comment by jeenbroekstra** [comment]
> Just to comment on setting the jvm in the benchmarks. I found we needed to do this to be able to run the benchmarks. It wasn't very portable no, but it was the only way I could find to make it work.

Would that still be the case with new set-up  though?  I had the impression it was specifically something to do with the way the elasticsearch cluster was spun up, and  that's now changed completely.

**Comment by barthanssens** [comment]
> I run the benchmarks directly from IntelliJ with a JMH plugin someone made.

@hmottestad  I'm a total IntelliJ newbie, but would this new plugin help (appears to allow for more settings) ? https://plugins.jetbrains.com/plugin/13868-jmhack

**Comment by hmottestad** [comment]
> > I run the benchmarks directly from IntelliJ with a JMH plugin someone made.
> 
> @hmottestad I'm a total IntelliJ newbie, but would this new plugin help (appears to allow for more settings) ? https://plugins.jetbrains.com/plugin/13868-jmhack

That looks interesting. I will try it, but not until maybe next week (on holiday now and quite tired after a 5 hour hike in the ⛰).

**Comment by jeenbroekstra** [comment]
Fwiw I've tried the new benchmark setup locally from Eclipse and it works fine.

We should make the jmh benchmarks a little easier to use though:

1. make sure all benchmark classes have a `main` method so they can be executed independently;
2. make sure all modules involving microbenchmarks produce a shaded jar, so that benchmarks can be executed from the command line (see https://mkyong.com/java/java-jmh-benchmark-tutorial/).

Anyway, both things are out of scope for this PR - I may tinker a little bit locally though and see if I can get a consistent setup that works from the command line and/or IDE without the need for a special plugin.

**Comment by hmottestad** [comment]
Did you intentionally close this @jeenbroekstra ?

**Comment by jeenbroekstra** [comment]
Explaining here as well for clarity: this got closed accidentally. When I merged develop into master for the 3.4 release, the develop branch automatically got deleted. I restored it, but I think Github automatically closes any pull request that targets a closed branch.

There may be a few other PRs that we'll need to restore.

File: core/sail/elasticsearch-store/src/test/java/org/eclipse/rdf4j/sail/elasticsearchstore/benchmark/AddBenchmark.java
Comment (barthanssens) [reply core/sail/elasticsearch-store/src/test/java/org/eclipse/rdf4j/sail/elasticsearchstore/benchmark/AddBenchmark.java]: ok, done

File: core/sail/elasticsearch-store/src/test/java/org/eclipse/rdf4j/sail/elasticsearchstore/TestHelpers.java
Comment (barthanssens) [reply core/sail/elasticsearch-store/src/test/java/org/eclipse/rdf4j/sail/elasticsearchstore/TestHelpers.java]: Removed the (test)method

**Review by jeenbroekstra — APPROVED** [review]
From a code perspective this all looks fine to me. Where are we with the CQs? I notice on the issue page there are still a few unticked.

**Comment by barthanssens** [comment]
Indeed, there are a few CQs waiting for approval, and I still need to file a few minor ones. I'll send a reminder to the IP team

The remaining issue is: compatibility testing. As in:

- using new Lucene version on existing Sail
- using new ES version on existing Sail
- connecting new Solr to older Solr server
- connecting new Elastic to older Elastic server

Lucene has a command-line tool to reindex existing indexes/stores, but I'm not sure about elasticsearch / Solr though.

I assume we'll need some additional tools (e.g testcontainers.org docker containers) to play with different versions.
But open to suggestions :-)

**Comment by jeenbroekstra** [comment]
@barthanssens I'm not sure I will have time but are you ok with me pushing commits to this branch if I find some time to look into it? I'd really like to get this over the line and out of the way.

**Comment by hmottestad** [comment]
Somewhat related to this. What are we going to do about the change in license for Elasticsearch?  I see that the Elasticsearch client has started using their new copy-left license (Server Side Public License) it seems. https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/main/java/org/elasticsearch/client/core/AcknowledgedResponse.java

I don't think it's applicable for the version we are upgrading to here, but it will be for future upgrades.

**Comment by barthanssens** [comment]
> Somewhat related to this. What are we going to do about the change in license for Elasticsearch? I see that the Elasticsearch client has started using their new copy-left license (Server Side Public License) it seems. 
> I don't think it's applicable for the version we are upgrading to here, but it will be for future upgrades.

Some more info: 
- https://www.elastic.co/pricing/faq/licensing
- https://www.elastic.co/licensing/elastic-license/faq

That would indeed be an issue for newer versions, looks like 7.10.x is the last Apache-licensed version...

**Comment by jeenbroekstra** [comment]
> Somewhat related to this. What are we going to do about the change in license for Elasticsearch? I see that the Elasticsearch client has started using their new copy-left license (Server Side Public License) it seems. https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/src/main/java/org/elasticsearch/client/core/AcknowledgedResponse.java
> 
> I don't think it's applicable for the version we are upgrading to here, but it will be for future upgrades.

Thanks for pointing this out, I wasn't aware of it. I haven't looked in detail yet but we will need to check how our usage fits in with their terms, and whether we have any additional obligations. I also want to be sure that there's no adverse effects to our users. We'll likely need to involve the IP team at some point to have a look as well.

**Comment by jeenbroekstra** [comment]
Fwiw I am currently getting the same test failure locally. This is in the Elasticsearch Store, rather than the FTS Sail.

**Comment by jeenbroekstra** [comment]
Problem might be related to the failsafe plugin using a forked runner - I had to disable this for most compliance tests when running in Java 9 or higher. I disabled it locally and got a green light, so pushed a fix. 

(fwiw I would not mind understanding why this happens, but for now I'm fine with just disabling forking)

**Comment by jeenbroekstra** [comment]
@barthanssens that seems to have fixed the failing test. Is there anything else you want to test before we merge this?

**Comment by barthanssens** [comment]
@jeenbroekstra thanks for the help !

For the Lucene part, I've added a check to test if an index was written with an older version (and if so, run a - potentially long - reindexing)
Hopefully this will also work for Solr (since that's based on Lucene)

So it is 'only' the ES store and the ES search, so yeah, still some testing...

**Review by jeenbroekstra — APPROVED** [review]
Codewise this looks fine to me as-is, thanks for all the wrangling on this @barthanssens ! 

I'm happy for this to be merged and called done. Could you perhaps look at the new release notes branch I've started and add a few relevant pointers for this change to the upgrade notes section?

**Comment by barthanssens** [comment]
Ok, will do, there are indeed some "dont-forget-to-reindex" notes to be written :-)

**Pull request merged**
